toml/src/main/kotlin/org/rust/toml/completion/crates/local/CratesLocalIndexService.kt

```diff
@@ -0,0 +1,212 @@
+        get() = baseCratesLocalRegistryDir.resolve(CORRUPTION_MARKER_NAME)
+
+    private val userCargoIndexDir: Path
+        get() = Paths.get(
+            System.getProperty("user.home"),
+            ".cargo/registry/index/github.com-1ecc6299db9ec823/.git/"
+        )
+
+    private val repository: Repository = FileRepositoryBuilder()
+        .setGitDir(userCargoIndexDir.toFile())
+        .build()
+
+    private var crates: PersistentHashMap<String, CargoRegistryCrate>
+
+    private val state: CratesLocalIndexState = service<CratesLocalIndexState>()
+
+    private var updateIsNeeded: Boolean = true
+
+    @Volatile
+    var isReady: Boolean = true
+
+    init {
+        checkInvalidatedStorage()
+
+        crates = PersistentHashMap(
+            baseCratesLocalRegistryDir.resolve("crates-local-index"),
+            EnumeratorStringDescriptor.INSTANCE,
+            CrateExternalizer
+        )
+
+        checkIfNeedUpdate()
+    }
+
+    private fun checkIfNeedUpdate() {
+        val branch = repository.resolve("origin/master") ?: return
+        val headCommitHash = branch.name
+
+        if (headCommitHash == state.indexedCommitHash) {
+            updateIsNeeded = false
+        }
+    }
+
+    private fun updateIfNeeded() {
+        if (updateIsNeeded) {
+            updateIsNeeded = false
+            isReady = false
+            CratesLocalIndexUpdateTask().queue()
+        }
+    }
+
+    fun checkInvalidatedStorage() {
+        if (corruptionMarkerFile.exists()) {
+            baseCratesLocalRegistryDir.cleanDirectory()
+            state.indexedCommitHash = ""
+        }
+    }
+
+    fun getCrate(crateName: String): CargoRegistryCrate? {
+        updateIfNeeded()
+        return crates.get(crateName)
+    }
+
+    fun getAllCrateNames(): List<String> {
+        updateIfNeeded()
+
+        val crateNames = mutableListOf<String>()
+        crates.processKeys { name ->
+            crateNames.add(name)
+        }
+
+        return crateNames
+    }
+
+    fun reloadCrates(revTree: RevTree) {
+        val gson = Gson()
+
+        TreeWalk(repository).use { treeWalk ->
+            treeWalk.addTree(revTree)
+            treeWalk.isRecursive = true
+            treeWalk.isPostOrderTraversal = false
+
+            while (treeWalk.next()) {
+                if (treeWalk.nameString == "config.json") continue
+
+                val objectId = treeWalk.getObjectId(0)
+                val loader = repository.open(objectId)
+                val versions = mutableListOf<CargoRegistryCrateVersion>()
+                val reader = BufferedReader(InputStreamReader(loader.openStream()))
+
+                reader.forEachLine { line ->
+                    if (line.isBlank()) return@forEachLine
+
+                    try {
+                        val parsedVersion = gson.fromJson(line, ParsedVersion::class.java)
+                        versions.add(CargoRegistryCrateVersion(parsedVersion.vers, parsedVersion.features.map { it.key }))
```

Comment:
Let's add `fun fromJson(json: String): CargoRegistryCrateVersion` to `CargoRegistryCrateVersion`'s companion object (and incapsulate `ParsedVersion` class inside the companion). Also, looks like `gson` used only once and can be inlined